### PR TITLE
gflow.d: double speed of fillInDNunambig()

### PR DESCRIPTION
**Pull request by WalterBright**
@ibuclaw  notes its quadratic performance here https://issues.dlang.org/show_bug.cgi?id=6401#c4

This change doubles its speed by noting that the matrix it creates is symmetric, so no need to redo half the work. But it's still quadratic.

The change:

1. splits off allocation of the DNunambig vectors from their computation
2. the inner loop starts at `start` rather than 0
3. removes extra checks as being redundant with the DNunambig vector being non-null

It isn't obvious that it is correct. So I added a check that computed the vectors the old way, and compared for identity with the new results. After it passes, this check can be deleted as being too expensive.

### CODE REVIEW

**Comment by WalterBright** [comment]
Hmm, it passes on my local machine.

**Comment by WalterBright** [comment]
Looks like adding that check was a good idea, it found two bugs in the new version.

**Comment by RazvanN7** [comment]
> splits off allocation of the DNunambig vectors from their computation

Why is this necessary? How does this help performance?

>  After it passes, this check can be deleted as being too expensive.

It's green now.

**Comment by ibuclaw** [comment]
> > splits off allocation of the DNunambig vectors from their computation
> 
> Why is this necessary? How does this help performance?
> 

Originally the nested foreach loop of `0..n` meant number of iterations it did was `(n*n)`, now the number of iterations would be around `(n*n/2)`.

File: src/dmd/backend/gflow.d [hunk 1]
```diff
@@ -291,6 +291,13 @@ private void asgdefelems(block *b,elem *n, DefNode[] defnod, ref size_t i)
     }
 }
 
+/* The easiest way to test this non-trivial revision is to compare it with the
+ * result from the previous implementation. (I already botched it once.)
+ * After it passes, remove.
+ */
+version = CheckAgainstOldMethod;
```
Comment (ibuclaw) [inline src/dmd/backend/gflow.d:298]: I guess this can be disabled now.

**Comment by WalterWaldron** [comment]
I've posted a rough draft of a more linear solution here:
https://github.com/dlang/dmd/compare/master...WalterWaldron:dnumambig

It doesn't solve extreme degenerate cases such as https://issues.dlang.org/show_bug.cgi?id=7157 but it should out do this solution.

**Comment by ibuclaw** [comment]
[Bench 1](https://github.com/dlang/dmd/pull/12456#issuecomment-824379934)  (Re-running because I'm getting different results now - I guess the background noise on my machine must be busier than usual)
| Baseline bfad3f04c311b8f82a415f89c0f3828661ad3341 | #12409 | #12453 | #12455 | #12456 | #12459 | @WalterWaldron  0ea9d777bd18689635ce0041ff9eb2905f62c53c | Baseline+WalterWaldron |
|-------------|------------|------------|-----------|------------|------------|------------|------------|
| 32.543s | 38.480s | 36.161s | 35.940s | 36.314s | 33.694s | 32.649s | 27.361s |

[Bench 2](https://issues.dlang.org/show_bug.cgi?id=6401#c6)

| Baseline bfad3f04c311b8f82a415f89c0f3828661ad3341 | #12409 | #12453 | #12455 | #12456 | #12459 | @WalterWaldron 0ea9d777bd18689635ce0041ff9eb2905f62c53c | Baseline+WalterWaldron |
|-------------|------------|------------|-----------|------------|------------|------------|------------|
| 1m23.593s | 1m34.041s | 1m32.060s | 1m32.213s | 1m32.993s | 1m24.946s | 1m22.961s | 1m17.279s | 1m17.279s |

At best, it's a marginal gain.  Should really profile it to drill down a bit more on the lowest hanging fruit.  As fixing this function isn't giving a satisfying boost, if the goal is to be as fast as gdc (7.132s and 14.229s respectively)

**Comment by WalterWaldron** [comment]
> At best, it's a marginal gain. Should really profile it to drill down a bit more on the lowest hanging fruit. As fixing this function isn't giving a satisfying boost, if the goal is to be as fast as gdc.

True. I'm planning to tackle the other O(N^2) backend functions mentioned in https://issues.dlang.org/show_bug.cgi?id=6401#c4 :
- ```gother.d:accumda``` is simple to fix
- ```gflow.d:accumaecpx``` and ```gdag.d:aewalk``` are slightly more complicated
- The newly posted ```mscoffobj.d:MsCoffObj_getsegment``` is also a simple fix.

It all boils down to using hash tables instead of O(N) search for each of the problem functions.

**Comment by RazvanN7** [comment]
ping @WalterBright

**Comment by RazvanN7** [comment]
@WalterBright what are your plans with this PR?

**Comment by RazvanN7** [comment]
ping @WalterBright

**Comment by WalterBright** [comment]
I recommend pulling this and then when @WalterWaldron 's solution is ready, pull that one.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Thanks!

**Pull request merged**